operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java

```diff
@@ -350,6 +356,40 @@ private StatefulSet createBaseDeployment(Keycloak keycloakCR, Context<Keycloak>
         return baseDeployment;
     }
 
+    private void handleScheduling(Keycloak keycloakCR, Map<String, String> labels, PodSpecFluent<?> specBuilder) {
+        SchedulingSpec schedulingSpec = keycloakCR.getSpec().getSchedulingSpec();
+        if (schedulingSpec != null) {
+            if (!specBuilder.hasPriorityClassName()) {
+                specBuilder.withPriorityClassName(schedulingSpec.getPriorityClassName());
+            }
+            if (!specBuilder.hasAffinity()) {
+                specBuilder.withAffinity(schedulingSpec.getAffinity());
+            }
+            if (!specBuilder.hasTolerations()) {
+                specBuilder.withTolerations(schedulingSpec.getTolerations());
+            }
+            if (!specBuilder.hasTopologySpreadConstraints()) {
+                specBuilder.withTopologySpreadConstraints(schedulingSpec.getTopologySpreadConstraints());
+            }
+        }
+
+        // set defaults if nothing was specified by the user
+
+        if (!specBuilder.hasAffinity()) {
+            specBuilder.editOrNewAffinity().withNewPodAffinity().addNewPreferredDuringSchedulingIgnoredDuringExecution()
+                    .withWeight(1).withNewPodAffinityTerm().withNewLabelSelector().withMatchLabels(labels)
```

Comment:
Shouldn't we be setting a greater weight here? The default policy should encourage pods to be distributed to Nodes in the same AZ as much as possible.

Reply:
From what we've talked about for default support this is sufficient. If we also add the ability to retain the default affinities, then we can make this greater to allow for users to choose weights that are less. 

In the worst case if there's just a single node in each AZ my expectation is that we'll still prefer separate nodes over keeping things in the same AZ - do you want it the other way around?

Reply:
It's a trade-off between performance and protecting against data loss. As this is the default policy, I think you're right that in the worst-case we should favour distributing pods across AZs in  order to prevent data loss on node failure.

However with a weight of 1 for the ZONE_KEY, what would happen in this scenario: K8s is deployed across 3 AZs, each of which has 2 nodes. Keycloak is then deployed with 3 replicas? Does having a lower weight assigned to the ZONE_KEY affinity strategy mean it's more likely that the 3rd pod will be deployed in a different AZ to the first 2?

Reply:
> Does having a lower weight assigned to the ZONE_KEY affinity strategy mean it's more likely that the 3rd pod will be deployed in a different AZ to the first 2 that are hosted on distinct nodes in the same AZ?

Let's assume that the first instance is placed in AZ 1. The next instance has 6 nodes to consider, kubernetes will create a sum of the weights for each of the targets. The same node will have a weight of 50, the other node in the same AZ will have a weight of 0. All of the other nodes will have a weight of 0. The scheduler will choose the one with the lowest weight, which will be the other node in the same AZ.

Reply:
Not meaning to nitpick, just want to clarify for my own understanding.

> The same node will have a weight of 50, the other node in the same AZ will have a weight of 0.

Won't the other node have a weight of 1 as it satisfies the zone affinity rule?


> The scheduler will choose the one with the lowest weight, which will be the other node in the same AZ.

I thought the scheduler chose the node with the largest value?


```qoute
The final sum is added to the score of other priority functions for the node. Nodes with the highest total score are prioritized when the scheduler makes a scheduling decision for the Pod.
```

Reply:
@ryanemerson yes sorry I was thinking of it the other way around for some reason. So what we actually have is:

node 1 zone 1 with an existing keycloak, score 1
node 2 zone 1 with score 51
node 3 zone 2 with score 50
node 4 zone 2 with score 50
node 5 zone 3 with score 50
node 6 zone 3 with score 50

Which should prefer the placemtn on node 2.

The more constrained scenario of just a single node per zone:

node 1 zone 1 with an existing keycloak, score 1
node 2 zone 2 with score 50
node 3 zone 3 with score 50

The scheduler will prefer node 2 / 3 over node 1.

Reply:
Thanks for the explanation @shawkins, it makes it clear to me now that a larger weight isn't required. The defaults LGTM :+1:

Reply:
In case we do move forward with an option to preserve the default affinities, I updated the PR to use 10 instead of 1 as the weight so that there is room for even smaller weights.